src/AbstractProduct.java

```diff
@@ -0,0 +1,34 @@
+import java.util.ArrayList;
+
+public abstract class AbstractProduct implements IProduct {
+  protected ArrayList<MaterialStack> payloadProduct = new ArrayList<>();
+  protected ArrayList<MaterialStack> costProduct = new ArrayList<>();
+  protected String category;

```

Comment:
no need to remember the category inside the product. All you have to do is to create a hashmap in the Shop instead so it knows which product is from what category.
Think of Category of being a Shop "room" or "isle", where they sell Milk stuff for example. The "Cheese" product does not have written on it "I am located in milk section". instead, it is just located in milk section - and milk section belongs to shop. If it makes sense in real world, then it will also make sense in OOP